File: 2216-delete-the-middle-node-of-a-linked-list/solution.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function deleteMiddle(head: ListNode | null): ListNode | null {
  let n = 0;
  let now = head;
  while (now) (now = now.next), n++;
  let pos = Math.floor(n / 2);
  if (pos === 0) return null;

  n = 0;
  now = head;
  while (n < pos - 1) (now = (now as ListNode).next), n++;
  (now as ListNode).next = now?.next?.next ?? null;
  return head;
}

File: 2216-delete-the-middle-node-of-a-linked-list/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* deleteMiddle(ListNode* head) {
        int n = 0;
        ListNode* node = head;
        while(node != NULL) n++, node = node->next;
        node = head;
        if(n == 1) return NULL;
        int idx = 0;
        while(idx < n / 2 - 1) {
            idx++;
            node = node->next;
        }
        ListNode* nextNext = node->next->next;
        node->next = nextNext;
        return head;
    }
};
